File: project-tab/adapter-shim/claude/src/mock-runner.ts
/**
 * Mock agent runner that emits a scripted sequence of events.
 *
 * The scripted sequence is:
 *   1. LifecycleEvent(started)
 *   2. StatusEvent("Analyzing codebase...")
 *   3. ToolCallEvent sequence (Read tool — reading a file)
 *   4. StatusEvent("Planning implementation...")
 *   5. ToolCallEvent sequence (Edit tool — modifying code)
 *   6. OptionDecisionEvent — agent proposes design choices, WAITS for resolution
 *   7. After resolve: ArtifactEvent (code file produced)
 *   8. CompletionEvent(success)
 */

import { v4 as uuidv4 } from 'uuid'
import { EventFactory } from './events.js'
import type {
  AdapterEvent,
  AgentBrief,
  AgentEvent,
  AgentHandle,
  KillResponse,
  ResolveRequest,
  SerializedAgentState,
} from './models.js'

export class MockRunner {
  static readonly PLUGIN_NAME = 'claude-mock'

  readonly brief: AgentBrief
  readonly agentId: string
  readonly sessionId: string

  private _factory: EventFactory
  private _handle: AgentHandle
  private _eventBuffer: AdapterEvent[] = []
  private _pendingDecisionId: string | null = null
  private _killed = false
  private _completed = false
  private _running = false

  // For decision blocking: resolve callback
  private _decisionResolve: ((req: ResolveRequest) => void) | null = null
  private _decisionReject: ((reason?: unknown) => void) | null = null
  private _runPromise: Promise<void> | null = null

  constructor(brief: AgentBrief) {
    this.brief = brief
    this.agentId = brief.agentId
    this.sessionId = uuidv4()
    this._factory = new EventFactory(uuidv4())
    this._handle = {
      id: this.agentId,
      pluginName: MockRunner.PLUGIN_NAME,
      status: 'running',
      sessionId: this.sessionId,
    }
  }

  get handle(): AgentHandle {
    return this._handle
  }

  get isRunning(): boolean {
    return this._running && !this._killed && !this._completed
  }

  get lastSequence(): number {
    return this._factory.lastSequence
  }

  start(): void {
    this._running = true
    this._runPromise = this._runScript()
  }

  private _emit(event: AgentEvent): void {
    const adapterEvent = this._factory.wrap(event)
    this._eventBuffer.push(adapterEvent)
  }

  private _sleep(ms: number): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this._killed) {
        resolve()
        return
      }
      const timer = setTimeout(resolve, ms)
      // Store a cleanup reference in case we get killed
      const checkKilled = setInterval(() => {
        if (this._killed) {
          clearTimeout(timer)
          clearInterval(checkKilled)
          resolve()
        }
      }, 10)
      // Auto-cleanup the interval when the timer fires naturally
      const origResolve = resolve
      // Note: the timer was already set, so we just clean up the interval on resolve
      void Promise.resolve().then(() => {
        // This won't affect anything, interval is cleaned by the check above
      })
    })
  }

  private async _runScript(): Promise<void> {
    try {
      // Step 1: LifecycleEvent(started)
      this._emit({
        type: 'lifecycle',
        agentId: this.agentId,
        action: 'started',
      })
      await this._sleep(50)
      if (this._killed) return

      // Step 2: StatusEvent
      this._emit({
        type: 'status',
        agentId: this.agentId,
        message: 'Analyzing codebase...',
      })
      await this._sleep(50)
      if (this._killed) return

      // Step 3: ToolCallEvent sequence (Read tool)
      const readToolCallId = uuidv4()
      this._emit({
        type: 'tool_call',
        agentId: this.agentId,
        toolCallId: readToolCallId,
        toolName: 'Read',
        phase: 'requested',
        input: { file_path: '/workspace/src/main.ts' },
        approved: true,
      })
      await this._sleep(30)

      this._emit({
        type: 'tool_call',
        agentId: this.agentId,
        toolCallId: readToolCallId,
        toolName: 'Read',
        phase: 'running',
        input: { file_path: '/workspace/src/main.ts' },
        approved: true,
      })
      await this._sleep(60)

      this._emit({
        type: 'tool_call',
        agentId: this.agentId,
        toolCallId: readToolCallId,
        toolName: 'Read',
        phase: 'completed',
        input: { file_path: '/workspace/src/main.ts' },
        output: { content: 'export function main() { /* ... */ }' },
        approved: true,
        durationMs: 90,
      })
      await this._sleep(30)
      if (this._killed) return

      // Step 4: StatusEvent
      this._emit({
        type: 'status',
        agentId: this.agentId,
        message: 'Planning implementation...',
      })
      await this._sleep(50)
      if (this._killed) return

      // Step 5: ToolCallEvent sequence (Edit tool)
      const editToolCallId = uuidv4()
      this._emit({
        type: 'tool_call',
        agentId: this.agentId,
        toolCallId: editToolCallId,
        toolName: 'Edit',
        phase: 'requested',
        input: {
          file_path: '/workspace/src/main.ts',
          old_string: 'export function main() { /* ... */ }',
          new_string: 'export function main() { return runPipeline(); }',
        },
        approved: true,
      })
      await this._sleep(30)

      this._emit({
        type: 'tool_call',
        agentId: this.agentId,
        toolCallId: editToolCallId,
        toolName: 'Edit',
        phase: 'completed',
        input: {
          file_path: '/workspace/src/main.ts',
          old_string: 'export function main() { /* ... */ }',
          new_string: 'export function main() { return runPipeline(); }',
        },
        output: { success: true },
        approved: true,
        durationMs: 45,
      })
      await this._sleep(30)
      if (this._killed) return

      // Step 6: OptionDecisionEvent — agent proposes design choices
      const decisionId = uuidv4()
      this._pendingDecisionId = decisionId
      this._handle = {
        ...this._handle,
        status: 'waiting_on_human',
      }

      this._emit({
        type: 'decision',
        subtype: 'option',
        agentId: this.agentId,
        decisionId,
        title: 'Architecture pattern for pipeline module',
        summary: 'The pipeline module can be structured as either a chain-of-responsibility or an event-driven architecture. Each has different tradeoffs for testability and extensibility.',
        severity: 'medium',
        confidence: 0.75,
        blastRadius: 'medium',
        options: [
          {
            id: 'opt-chain',
            label: 'Chain of Responsibility',
            description: 'Sequential pipeline stages with explicit handoff between steps.',
            tradeoffs: 'Simpler to debug and test, but harder to add parallel stages later.',
          },
          {
            id: 'opt-events',
            label: 'Event-Driven Architecture',
            description: 'Decoupled stages communicating through an event bus.',
            tradeoffs: 'More flexible and extensible, but harder to trace execution flow.',
          },
          {
            id: 'opt-hybrid',
            label: 'Hybrid Approach',
            description: 'Core pipeline is sequential, but supports event hooks for cross-cutting concerns.',
            tradeoffs: 'Balanced approach, moderate complexity.',
          },
        ],
        recommendedOptionId: 'opt-hybrid',
        requiresRationale: true,
      })

      // Wait for resolution
      await new Promise<ResolveRequest>((resolve, reject) => {
        this._decisionResolve = resolve
        this._decisionReject = reject
      })

      this._pendingDecisionId = null
      this._decisionResolve = null
      this._decisionReject = null

      if (this._killed) return

      // Step 7: After resolve — ArtifactEvent
      this._handle = {
        ...this._handle,
        status: 'running',
      }

      const artifactId = uuidv4()
      const nowIso = new Date().toISOString()
      this._emit({
        type: 'artifact',
        agentId: this.agentId,
        artifactId,
        name: 'pipeline.ts',
        kind: 'code',
        workstream: this.brief.workstream,
        status: 'draft',
        qualityScore: 0.85,
        provenance: {
          createdBy: this.agentId,
          createdAt: nowIso,
        },
        uri: '/workspace/src/pipeline.ts',
        mimeType: 'text/typescript',
        sizeBytes: 2048,
      })
      await this._sleep(30)
      if (this._killed) return

      // Step 8: CompletionEvent
      this._emit({
        type: 'completion',
        agentId: this.agentId,
        summary: 'Implemented pipeline module using the chosen architecture pattern. Created pipeline.ts with core pipeline logic.',
        artifactsProduced: [artifactId],
        decisionsNeeded: [],
        outcome: 'success',
      })

      this._handle = {
        ...this._handle,
        status: 'completed',
      }
      this._completed = true
    } catch (err) {
      // Killed or cancelled
      if (!this._killed) {
        this._emit({
          type: 'error',
          agentId: this.agentId,
          severity: 'high',
          message: err instanceof Error ? err.message : String(err),
          recoverable: false,
          category: 'internal',
        })
      }
    }
  }

  resolveDecision(request: ResolveRequest): boolean {
    if (
      this._decisionResolve !== null &&
      request.decisionId === this._pendingDecisionId
    ) {
      this._decisionResolve(request)
      return true
    }
    return false
  }

  async kill(grace: boolean = true): Promise<KillResponse> {
    this._killed = true
    if (this._decisionReject) {
      this._decisionReject(new Error('killed'))
      this._decisionResolve = null
      this._decisionReject = null
    }

    // Wait briefly for run script to exit
    if (this._runPromise) {
      await Promise.race([
        this._runPromise.catch(() => {}),
        new Promise(r => setTimeout(r, 100)),
      ])
    }

    this._emit({
      type: 'lifecycle',
      agentId: this.agentId,
      action: 'killed',
      reason: `kill requested${grace ? ' (graceful)' : ' (force)'}`,
    })

    this._handle = {
      ...this._handle,
      status: 'completed',
    }
    this._running = false

    return {
      state: null,
      artifactsExtracted: 0,
      cleanShutdown: grace,
    }
  }

  async pause(): Promise<SerializedAgentState> {
    this._killed = true
    if (this._decisionReject) {
      this._decisionReject(new Error('paused'))
      this._decisionResolve = null
      this._decisionReject = null
    }

    if (this._runPromise) {
      await Promise.race([
        this._runPromise.catch(() => {}),
        new Promise(r => setTimeout(r, 100)),
      ])
    }

    this._emit({
      type: 'lifecycle',
      agentId: this.agentId,
      action: 'paused',
    })

    this._handle = {
      ...this._handle,
      status: 'paused',
    }
    this._running = false

    const nowIso = new Date().toISOString()
    return {
      agentId: this.agentId,
      pluginName: MockRunner.PLUGIN_NAME,
      sessionId: this.sessionId,
      checkpoint: {
        sdk: 'claude-mock',
        scriptPosition: this._factory.lastSequence,
      },
      briefSnapshot: this.brief,
      pendingDecisionIds: this._pendingDecisionId ? [this._pendingDecisionId] : [],
      lastSequence: this._factory.lastSequence,
      serializedAt: nowIso,
      serializedBy: 'pause',
      estimatedSizeBytes: 256,
    }
  }

  drainEvents(): AdapterEvent[] {
    const events = [...this._eventBuffer]
    this._eventBuffer = []
    return events
  }
}

File: project-tab/adapter-shim/openai/adapter_shim/mock_runner.py
"""Mock agent runner that emits a scripted sequence of events.

The scripted sequence is:
  1. LifecycleEvent(started)
  2. StatusEvent("Starting task...")
  3. ToolCallEvent(requested) -> ToolCallEvent(running) -> ToolCallEvent(completed)
  4. DecisionEvent(tool_approval) -- then WAIT for POST /resolve
  5. After resolve: ArtifactEvent
  6. CompletionEvent(success)
"""

from __future__ import annotations

import asyncio
import uuid
from datetime import datetime, timezone
from typing import TYPE_CHECKING

from .events import EventFactory
from .models import (
    AgentBrief,
    AgentHandle,
    ArtifactEvent,
    CompletionEvent,
    KillResponse,
    LifecycleEvent,
    Provenance,
    ResolveRequest,
    SerializedAgentState,
    StatusEvent,
    ToolApprovalEvent,
    ToolCallEvent,
)

if TYPE_CHECKING:
    from .models import AdapterEvent


class MockRunner:
    """Runs a scripted mock agent that emits events over time."""

    PLUGIN_NAME = "openai-mock"

    def __init__(self, brief: AgentBrief) -> None:
        self.brief = brief
        self.agent_id = brief.agent_id
        self.session_id = str(uuid.uuid4())
        self._factory = EventFactory(run_id=str(uuid.uuid4()))
        self._status: AgentHandle = AgentHandle(
            id=self.agent_id,
            pluginName=self.PLUGIN_NAME,
            status="running",
            sessionId=self.session_id,
        )
        self._event_buffer: list[AdapterEvent] = []
        self._ws_connected = False
        self._decision_future: asyncio.Future[ResolveRequest] | None = None
        self._pending_decision_id: str | None = None
        self._task: asyncio.Task[None] | None = None
        self._killed = False
        self._completed = False

    @property
    def handle(self) -> AgentHandle:
        return self._status

    @property
    def is_running(self) -> bool:
        return not self._killed and not self._completed

    def start(self) -> None:
        """Start the mock event sequence in the background."""
        self._task = asyncio.create_task(self._run_script())

    async def _emit(self, event: object) -> None:
        """Wrap and buffer an event."""
        adapter_event = self._factory.wrap(event)  # type: ignore[arg-type]
        self._event_buffer.append(adapter_event)

    async def _run_script(self) -> None:
        """Execute the scripted event sequence."""
        try:
            # Step 1: LifecycleEvent(started)
            await self._emit(LifecycleEvent(
                agentId=self.agent_id,
                action="started",
            ))
            await asyncio.sleep(0.1)

            if self._killed:
                return

            # Step 2: StatusEvent
            await self._emit(StatusEvent(
                agentId=self.agent_id,
                message="Starting task...",
            ))
            await asyncio.sleep(0.1)

            if self._killed:
                return

            # Step 3: ToolCallEvent sequence
            tool_call_id = str(uuid.uuid4())
            await self._emit(ToolCallEvent(
                agentId=self.agent_id,
                toolCallId=tool_call_id,
                toolName="file_search",
                phase="requested",
                input={"query": "project requirements"},
                approved=True,
            ))
            await asyncio.sleep(0.05)

            await self._emit(ToolCallEvent(
                agentId=self.agent_id,
                toolCallId=tool_call_id,
                toolName="file_search",
                phase="running",
                input={"query": "project requirements"},
                approved=True,
            ))
            await asyncio.sleep(0.1)

            await self._emit(ToolCallEvent(
                agentId=self.agent_id,
                toolCallId=tool_call_id,
                toolName="file_search",
                phase="completed",
                input={"query": "project requirements"},
                output={"results": ["requirements.md"]},
                approved=True,
                durationMs=150,
            ))
            await asyncio.sleep(0.05)

            if self._killed:
                return

            # Step 4: DecisionEvent(tool_approval) -- wait for resolve
            decision_id = str(uuid.uuid4())
            self._pending_decision_id = decision_id
            self._status = AgentHandle(
                id=self.agent_id,
                pluginName=self.PLUGIN_NAME,
                status="waiting_on_human",
                sessionId=self.session_id,
            )

            await self._emit(ToolApprovalEvent(
                agentId=self.agent_id,
                decisionId=decision_id,
                toolName="execute_code",
                toolArgs={"code": "print('hello world')", "language": "python"},
                severity="medium",
                confidence=0.85,
                blastRadius="small",
            ))

            # Create a future and wait for resolution
            loop = asyncio.get_running_loop()
            self._decision_future = loop.create_future()
            try:
                await self._decision_future
            except asyncio.CancelledError:
                return

            self._pending_decision_id = None
            self._decision_future = None

            if self._killed:
                return

            # Step 5: After resolve -- ArtifactEvent
            self._status = AgentHandle(
                id=self.agent_id,
                pluginName=self.PLUGIN_NAME,
                status="running",
                sessionId=self.session_id,
            )

            now_iso = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
            artifact_id = str(uuid.uuid4())
            await self._emit(ArtifactEvent(
                agentId=self.agent_id,
                artifactId=artifact_id,
                name="report.md",
                kind="document",
                workstream=self.brief.workstream,
                status="draft",
                qualityScore=0.9,
                provenance=Provenance(
                    createdBy=self.agent_id,
                    createdAt=now_iso,
                ),
                uri="/workspace/output/report.md",
                mimeType="text/markdown",
                sizeBytes=1024,
            ))
            await asyncio.sleep(0.05)

            if self._killed:
                return

            # Step 6: CompletionEvent(success)
            await self._emit(CompletionEvent(
                agentId=self.agent_id,
                summary="Mock task completed successfully. Generated report.md.",
                artifactsProduced=[artifact_id],
                decisionsNeeded=[],
                outcome="success",
            ))

            self._status = AgentHandle(
                id=self.agent_id,
                pluginName=self.PLUGIN_NAME,
                status="completed",
                sessionId=self.session_id,
            )
            self._completed = True

        except asyncio.CancelledError:
            pass

    def resolve_decision(self, request: ResolveRequest) -> bool:
        """Resolve a pending decision. Returns True if resolved."""
        if (
            self._decision_future is not None
            and not self._decision_future.done()
            and request.decision_id == self._pending_decision_id
        ):
            self._decision_future.set_result(request)
            return True
        return False

    async def kill(self, grace: bool = True) -> KillResponse:
        """Kill the mock agent."""
        self._killed = True
        if self._decision_future and not self._decision_future.done():
            self._decision_future.cancel()
        if self._task and not self._task.done():
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass

        await self._emit(LifecycleEvent(
            agentId=self.agent_id,
            action="killed",
            reason="kill requested" + (" (graceful)" if grace else " (force)"),
        ))

        self._status = AgentHandle(
            id=self.agent_id,
            pluginName=self.PLUGIN_NAME,
            status="completed",
            sessionId=self.session_id,
        )

        return KillResponse(
            state=None,
            artifactsExtracted=0,
            cleanShutdown=grace,
        )

    async def pause(self) -> SerializedAgentState:
        """Pause the mock agent and return serialized state."""
        self._killed = True
        if self._decision_future and not self._decision_future.done():
            self._decision_future.cancel()
        if self._task and not self._task.done():
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass

        await self._emit(LifecycleEvent(
            agentId=self.agent_id,
            action="paused",
        ))

        self._status = AgentHandle(
            id=self.agent_id,
            pluginName=self.PLUGIN_NAME,
            status="paused",
            sessionId=self.session_id,
        )

        now_iso = datetime.now(timezone.utc).isoformat()
        from .models import SdkCheckpoint

        return SerializedAgentState(
            agentId=self.agent_id,
            pluginName=self.PLUGIN_NAME,
            sessionId=self.session_id,
            checkpoint=SdkCheckpoint(
                sdk="mock",
                scriptPosition=self._factory.last_sequence,
            ),
            briefSnapshot=self.brief,
            pendingDecisionIds=[self._pending_decision_id] if self._pending_decision_id else [],
            lastSequence=self._factory.last_sequence,
            serializedAt=now_iso,
            serializedBy="pause",
            estimatedSizeBytes=256,
        )

    def get_checkpoint(self, decision_id: str) -> SerializedAgentState:
        """Return a checkpoint snapshot without stopping the agent."""
        now_iso = datetime.now(timezone.utc).isoformat()
        from .models import SdkCheckpoint

        return SerializedAgentState(
            agentId=self.agent_id,
            pluginName=self.PLUGIN_NAME,
            sessionId=self.session_id,
            checkpoint=SdkCheckpoint(
                sdk="mock",
                scriptPosition=self._factory.last_sequence,
            ),
            briefSnapshot=self.brief,
            conversationSummary="Agent blocked on decision",
            pendingDecisionIds=[decision_id] if decision_id else [],
            lastSequence=self._factory.last_sequence,
            serializedAt=now_iso,
            serializedBy="decision_checkpoint",
            estimatedSizeBytes=256,
        )

    def drain_events(self) -> list[AdapterEvent]:
        """Drain all buffered events."""
        events = list(self._event_buffer)
        self._event_buffer.clear()
        return events
